src/Microsoft.DotNet.Cli.Utils/CommandResolution/PackagedCommandSpecCreator.cs

```diff
@@ -0,0 +1,108 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using Microsoft.DotNet.ProjectModel;
+using Microsoft.DotNet.ProjectModel.Graph;
+using Microsoft.Extensions.PlatformAbstractions;
+using NuGet.Frameworks;
+using NuGet.Packaging;
+
+namespace Microsoft.DotNet.Cli.Utils
+{
+    internal class PackagedCommandSpecCreator : IPackagedCommandSpecCreator
+    {
+        public CommandSpec CreateCommandSpecFromLibrary(
+            LockFilePackageLibrary library,
+            string commandName,
+            IEnumerable<string> commandArguments,
+            IEnumerable<string> allowedExtensions,
+            string nugetPackagesRoot,
+            string depsFilePath)
+        {
+            var packageDirectory = GetPackageDirectoryFullPath(library, nugetPackagesRoot);
+
+            if (!Directory.Exists(packageDirectory))
+            {
+                return null;
+            }
+
+            var commandFile = GetCommandFileRelativePath(library, commandName, allowedExtensions);
+
+            if (commandFile == null)
+            {
+                return null;
+            }
+
+            var commandPath = Path.Combine(packageDirectory, commandFile);
+
+            return CreateCommandSpecUsingCoreHostIfDll(commandPath, commandArguments, depsFilePath);
+        }
+
+        private string GetPackageDirectoryFullPath(LockFilePackageLibrary library, string nugetPackagesRoot)
+        {
+            var packageDirectory = new VersionFolderPathResolver(nugetPackagesRoot)
+                .GetInstallPath(library.Name, library.Version);
+
+            return packageDirectory;
+        }
+
+        private string GetCommandFileRelativePath(
+            LockFilePackageLibrary library, 
+            string commandName, 
+            IEnumerable<string> allowedExtensions)
+        {
+            // TODO: Should command names be case sensitive?
+            return library.Files
+                    .Where(f => Path.GetFileNameWithoutExtension(f) == commandName)
+                    .Where(e => allowedExtensions.Contains(Path.GetExtension(e)))
+                    .FirstOrDefault();
+        }
+
+        private CommandSpec CreateCommandSpecUsingCoreHostIfDll(
```

Comment:
The name on this method implies that there is no alternative and it will only create if it is a dll. Which is not really the case.